### PR TITLE
Fix upgrade guide for v1.7 and replicate it in a CI test

### PR DESCRIPTION
**Pull request by aanm**
- Fix upgrade tests from v1.6 to master/v1.7.
- Re-write the upgrade

Depends on #10101 and #10102 

Signed-off-by: André Martins <andre@cilium.io>

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/9993)

### CODE REVIEW

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-focus K8sUpdates*

**Comment by aanm** [comment]
test-focus K8sUpdates*

**Comment by aanm** [comment]
test-focus K8sUpdates*

**Comment by aanm** [comment]
test-focus K8sUpdates*

**Comment by aanm** [comment]
test-focus K8sUpdates*

**Comment by aanm** [comment]
test-focus K8sUpdates*

**Comment by aanm** [comment]
test-focus K8sUpdates*

**Comment by aanm** [comment]
test-focus K8sUpdates*

**Comment by aanm** [comment]
test-focus K8sUpdates*

**Comment by aanm** [comment]
test-focus K8sUpdates*

**Comment by aanm** [comment]
test-focus K8sUpdates*

**Comment by aanm** [comment]
test-focus K8sUpdates*

**Comment by aanm** [comment]
test-focus K8sUpdates*

**Comment by aanm** [comment]
test-focus K8sUpdates*

**Comment by aanm** [comment]
test-focus K8sUpdates*

**Comment by aanm** [comment]
test-focus K8sUpdates*

**Comment by aanm** [comment]
test-focus K8sUpdates*

**Comment by aanm** [comment]
test-nightly K8sUpdates*

**Comment by aanm** [comment]
test-focus K8sUpdates*

**Comment by aanm** [comment]
test-focus K8sUpdates*

**Comment by aanm** [comment]
test-focus K8sUpdates*

**Comment by aanm** [comment]
test-focus K8sUpdates*

**Comment by aanm** [comment]
@raybejjani I had to revert one commit because it was causing the updates test to fail. Also, PTAL at the way I'm installing helm and running its sub-commands.

@joestringer WRT the `v1.6` I gave it a little thought, maybe we could have a `v1.6-dev` helm chart version in the helm repository. From what I've read helm does not pick dev versions unless the user specifies it.

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-nightly

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

File: Documentation/install/upgrade.rst [hunk 1]
```diff
@@ -309,22 +309,39 @@ IMPORTANT: Changes required before upgrading to 1.7.0
 * Cilium has bumped the minimal kubernetes version supported to v1.11.0.
 
 * The ``kubernetes.io/cluster-service`` label has been removed from the Cilium
-  `DaemonSet` selector. Existing users must remove this label from their
-  `DaemonSet` specification to safely upgrade. If action is not taken to make
-  the selector labels consistent with the upgraded YAMLs, then the new
-  `DaemonSet` YAML which is created during upgrade will fail to apply.
+  `DaemonSet` selector. Existing users must either choose to keep this label on
```
Comment (brb) [inline Documentation/install/upgrade.rst]: s/to keep this label on/to keep this label in/

File: Documentation/install/upgrade.rst [hunk 2]
```diff
@@ -309,22 +309,39 @@ IMPORTANT: Changes required before upgrading to 1.7.0
...
+
+  .. group-tab:: Helm
+
+    .. parsed-literal::
+
+      helm upgrade cilium --namespace=kube-system \
+      --set agent.keepDeprecatedLabels=true
+
+
+  Trying to upgrade Cilium without this option might result in an error similar
```
Comment (brb) [inline Documentation/install/upgrade.rst]: s/in an error similar to/in the following error/

File: Documentation/install/upgrade.rst [hunk 3]
```diff
@@ -309,22 +309,39 @@ IMPORTANT: Changes required before upgrading to 1.7.0
...
 * The ``kubernetes.io/cluster-service`` label has been removed from the Cilium
-  `DaemonSet` selector. Existing users must remove this label from their
-  `DaemonSet` specification to safely upgrade. If action is not taken to make
-  the selector labels consistent with the upgraded YAMLs, then the new
-  `DaemonSet` YAML which is created during upgrade will fail to apply.
+  `DaemonSet` selector. Existing users must either choose to keep this label on
+  `DaemonSet` specification to safely upgrade or re-create the Cilium `DaemonSet`
+  without the deprecated label. It is advisable to run the with this flag when
+  doing an upgrade from ``v1.6.x`` to ``v1.7.x`` in the event of having to do a
+  downgrade. Removing this label should be done within same Cilium versions.
```
Comment (brb) [inline Documentation/install/upgrade.rst]: Maybe "Removing the label should be done after the upgrade"?

File: Documentation/install/upgrade.rst [hunk 4]
```diff
@@ -309,22 +309,39 @@ IMPORTANT: Changes required before upgrading to 1.7.0
 * Cilium has bumped the minimal kubernetes version supported to v1.11.0.
 
 * The ``kubernetes.io/cluster-service`` label has been removed from the Cilium
-  `DaemonSet` selector. Existing users must remove this label from their
-  `DaemonSet` specification to safely upgrade. If action is not taken to make
-  the selector labels consistent with the upgraded YAMLs, then the new
-  `DaemonSet` YAML which is created during upgrade will fail to apply.
+  `DaemonSet` selector. Existing users must either choose to keep this label on
+  `DaemonSet` specification to safely upgrade or re-create the Cilium `DaemonSet`
+  without the deprecated label. It is advisable to run the with this flag when
```
Comment (brb) [inline Documentation/install/upgrade.rst]: "It is advisable to keep the label when doing <..>"?

File: test/helpers/kubectl.go [hunk 5]
```diff
@@ -1423,6 +1423,20 @@ func (kub *Kubectl) CiliumInstall(filename string, options map[string]string) er
 	return kub.ciliumInstallHelm(filename, options)
 }
 
+// CiliumHelm runs the helm command with the given options.
+func (kub *Kubectl) CiliumHelm(action, repo, helmName, version, namespace string, options map[string]string) *CmdRes {
```
Comment (brb) [inline test/helpers/kubectl.go]: s/CiliumHelm/RunHelm/?

File: test/helpers/cons.go [hunk 6]
```diff
@@ -174,9 +174,9 @@ const (
 	KubectlPolicyNameLabel      = k8sConst.PolicyLabelName
 	KubectlPolicyNameSpaceLabel = k8sConst.PolicyLabelNamespace
 
-	CiliumStableVersion      = "v1.5"
-	CiliumStableImageVersion = "cilium/cilium:" + CiliumStableVersion
-	ciliumDeveloperImage     = "%s/cilium/cilium-dev:latest"
+	CiliumStableHelmVersion = "1.6-dev"
```
Comment (brb) [inline test/helpers/cons.go]: A comment why `-dev`?

File: test/helpers/kubectl.go [hunk 7]
```diff
@@ -2884,6 +2898,11 @@ func (kub *Kubectl) reportMapHost(ctx context.Context, path string, reportCmds m
 	}
 }
 
+// HelmRepoInstall installs the repository that contain Cilium helm charts.
+func (kub *Kubectl) HelmRepoInstall() *CmdRes {
```
Comment (brb) [inline test/helpers/kubectl.go]: HelmAddCiliumRepo?

File: test/k8sT/Updates.go [hunk 8]
```diff
@@ -121,6 +133,37 @@ func InstallAndValidateCiliumUpgrades(kubectl *helpers.Kubectl, ciliumFilename,
 	apps := []string{helpers.App1, helpers.App2, helpers.App3}
 	app1Service := "app1-service"
 
+	ciliumCiliumState := func(helmPath, helmVersion string) {
+		_ = kubectl.ExecMiddle("helm delete cilium --namespace=" + helpers.CiliumNamespace)
+		_ = kubectl.ExecMiddle(fmt.Sprintf("kubectl delete configmap --namespace=%s cilium-config", helpers.CiliumNamespace))
```
Comment (brb) [inline test/k8sT/Updates.go:144]: Maybe move all `kubectl delete` instances into a separate function which could be later re-used in `AfterAll()` to remove all Cilium-related artifacts?

File: test/k8sT/Updates.go [hunk 9]
```diff
@@ -121,6 +133,37 @@ func InstallAndValidateCiliumUpgrades(kubectl *helpers.Kubectl, ciliumFilename,
 	apps := []string{helpers.App1, helpers.App2, helpers.App3}
 	app1Service := "app1-service"
 
+	ciliumCiliumState := func(helmPath, helmVersion string) {
```
Comment (brb) [inline test/k8sT/Updates.go]: s/ciliumCiliumState/cleanupCiliumState/

File: test/k8sT/Updates.go [hunk 10]
```diff
@@ -326,66 +352,73 @@ func InstallAndValidateCiliumUpgrades(kubectl *helpers.Kubectl, ciliumFilename,
...
-		ExpectWithOffset(1, err).To(BeNil(), "Cilium %q was not able to be deployed", oldVersion)
+		By("Downgrading cilium to %s image", oldHelmVersion)
+		// rollback cilium 1 because it's the version that we have started
+		// cilium with in this updates test.
+		cmd = kubectl.ExecMiddle("helm rollback cilium 1 --namespace=" + helpers.CiliumNamespace)
+		ExpectWithOffset(1, cmd).To(helpers.CMDSuccess(), "Cilium %q was not able to be deployed", oldHelmVersion)
 
 		err = helpers.WithTimeout(
-			waitForUpdateImage(oldVersion),
+			waitForUpdateImage(oldHelmVersion),
```
Comment (brb) [inline test/k8sT/Updates.go]: oldHelmVersion reads as a version of helm, but not a version of cilium.

**Review by brb — CHANGES_REQUESTED** [review]

File: test/k8sT/Updates.go (hunk 8 above)
Comment (aanm) [reply test/k8sT/Updates.go:144]: @brb it's already the case, this is called in line 103

File: Documentation/install/upgrade.rst (hunk 3 above)
Comment (aanm) [reply Documentation/install/upgrade.rst]: Change to `The removal of this label should be done after a successful upgrade.`

**Comment by aanm** [comment]
test-focus K8sUpdates*

File: Documentation/install/upgrade.rst [hunk 11]
```diff
@@ -309,22 +309,40 @@ IMPORTANT: Changes required before upgrading to 1.7.0
 * Cilium has bumped the minimal kubernetes version supported to v1.11.0.
 
 * The ``kubernetes.io/cluster-service`` label has been removed from the Cilium
-  `DaemonSet` selector. Existing users must remove this label from their
-  `DaemonSet` specification to safely upgrade. If action is not taken to make
-  the selector labels consistent with the upgraded YAMLs, then the new
-  `DaemonSet` YAML which is created during upgrade will fail to apply.
+  `DaemonSet` selector. Existing users must either choose to keep this label in
+  `DaemonSet` specification to safely upgrade or re-create the Cilium `DaemonSet`
+  without the deprecated label. It is advisable to keep the label when doing
```
Comment (brb) [inline Documentation/install/upgrade.rst:314]: Nit: s/doing doing/doing/

File: test/k8sT/Updates.go (hunk 8 above)
Comment (brb) [reply test/k8sT/Updates.go:144]: I meant in `AfterAll()` of other test suites. I.e. to move to the `test/helpers/`.

**Review by brb — CHANGES_REQUESTED** [review]

**Comment by aanm** [comment]
test-focus K8sUpdates*

**Review by brb — APPROVED** [review]

**Comment by aanm** [comment]
test-focus K8sUpdates*

**Comment by aanm** [comment]
test-focus K8sUpdates*

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-docs-please

**Comment by aanm** [comment]
test-focus K8sUpdates*

**Comment by aanm** [comment]
test-focus K8sUpdates*

**Comment by aanm** [comment]
test-focus K8sUpdates*

**Comment by aanm** [comment]
test-focus K8sUpdates*

**Comment by aanm** [comment]
test-me-please

**Pull request merged**